### DISCUSSION TITLE
py to exe tech

### DISCUSSION THREAD

**Discussion by dtoft** [pyinstaller/pyinstaller#8163] [PyInstaller]
What's happened with the state of py > exe tech? It's been a while (2017) since I've owned this process and, coming back, it seems like py2exe is no longer the leading open source package, at least in terms of support community. I used py2exe for a while with great results. But now it seems to have fallen behind python + package updates (doesn't work for my current need)

Now it seems like everyone is using PyInstaller, which is supposed to be better for the latest python. But after trying it for a few hours, I'm quickly disappointed. It doesn't build a professional distribution package like py2exe and seems to be a quick-fix for deploy-python-to-users-without-python (vs a more robust desktop app distribution). It doesn't handle app icons and Windows Properties Details well

So now I'm wondering if I should invest more time into getting py2exe to work and deal with what I expect to be ongoing issues compiling the data analysis packages. But it's got me wondering, what happened? What am I missing? Why did the community abandon py2exe? Why isn't PyInstaller building for professional deployment? Is there a different package I should be looking at?

Wondering what the community thinks!

**Comment by bwoodsend** [comment]
Define _professional distribution package_

**Reply by dtoft** [reply]
As in a distribution directory, staged for digital signing, installer build, to be distributed in a traditional desktop app deployment (download installer, run installer, run app). With a fair amount of control over how it's done. The DX and docs feel more like PyInstaller is more simply meant to solve getting python scripts to run without a python install, even if that means emailing a zipped package (because the traditional process gets complicated). Unfair?

FWIW, PyInstaller does seem to have a better recursive packaging and environment solver

**Reply by bwoodsend** [reply]
We feel that converting a relocatable bundle to an installer can already be achieved via inosetup and believe it's a waste of time to try and glue inosetup into PyInstaller given how little configuration they share.

**Reply by dtoft** [reply]
Sure, and Inno Setup was always part of the process with py2exe. Incidentally, Inno Setup has gotten better and added features over these 7 years. Ironically, this library is called PyInstaller, the branding suggests broader coverage than py2exe

And sure, we can explicitly build the dist directory (although not intuitive what it will break, what relative path dependencies are created from user POV) and Properties Details, etc with native packages. And we can back into the spec file args by running dummy tests with the documented command line args. These are all clever workarounds that minimize the work to make a python(ic) library

Tradeoff: using this package doesn't feel like using a python package to build an installer with python. To my original point and I think your point here, the goal of this project doesn't seem to be distributing python to desktop apps. It's more like a minimal, general-use py-to-exe converter (although the splash screen is an odd detour from this)

Reason I posted, think a community discussion is needed is that's what we had 7 years ago with py2exe. It's like the community pivoted to another project to work back to where we were (maybe my history's wrong?). Both projects have yin-yang strengths + weaknesses, neither is great on its own, and yet to me this is a critical part of the python ecosystem. Is there a discussion/project that wants to push the tech forward? If this project isn't interested in encroaching on Inno Setup's 90s DSL + DX or native python handwork, is there a viable project that is? Does it make sense for py2exe to focus on keeping it's pythonic DX working, using PyInstaller's engine? Does it make sense to build backwards with Inno Setup plugins? etc.

**Comment by albertosottile** [comment]
FTR, the same discussion was opened in the `py2exe` repository. I replied there with my two cents on this: https://github.com/py2exe/py2exe/discussions/197#discussioncomment-7835736
